### ISSUE TITLE
delay between OS bc trigger and jenkins build start based on order of namespaces configured

### ISSUE THREAD

**Issue by andreaseger** [openshift/jenkins-sync-plugin#227]
As mentioned in #200  I'm seeing a weird/interesting delay between triggering a pipeline in openshift and start of the execution in jenkins based on the order of namespaces I configure in jenkins. This only happens if I configure multiple namespaces to sync.

---

What I tried is having a namspace `jenkins` which has the slave image streams and a `playground` namespace where the jenkins and the buildconfig/pipeline is defined.

If I set in the Jenkins config "OpenShift Jenkins Sync" |> "Namespace" to "`jenkins playground`".

It finds the imagestreams from `jenkins` and adds the kubernetes pod templates. It also created a Pipeline project for the bc in `playground`. But if I use "start pipeline" in OS it seems like nothing is happening in jenkins - turns out I was not patient enough and after a few minutes the build would be started in jenkins. I can see a 5 minutes delay here.

If I change the order in the settings to "`playground jenkins`" then clicking on "start pipeline" in openshift almost immediately triggers the build in jenkins.

Can you explain why "`playground jenkins`" is nearly instant while "`jenkins playground`" has a huge delay?

**Comment by gabemontero** [comment]
hey @andreaseger - would you happen to be at v1.0.9 or later for the plugin?

We added logging in that version for when the watch is unexpectedly closed on us, as well as an attempt to re-establish the watch on the next list attempt.

Either watch(es) are getting closed on us, where maybe the second watch is getting closed first for some reason, and thus why your order matters.  Or there is something up with multiple namespace watches which we have not seen before.

Anyway, running with a sufficient version, get us the jenkins pod logs, and we'll debug from there.

**Comment by andreaseger** [comment]
we're running
jenkins `2.107.2`
jenkins-sync plugin `1.0.11`
openshift `v3.6.173.0.96`


I redeployed the jenkins pod and after it was up again I checked the config (here it's `jenkins ane-playground`) and triggered the builds from openshift. Took a few minutes till jenkins saw them.

There are a couple exceptions in the logs, not sure they are related.
[jenkins-11-4n163.log](https://github.com/openshift/jenkins-sync-plugin/files/1916028/jenkins-11-4n163.log)

**Comment by gabemontero** [comment]
OK the watch was not closed on us ... none of those logs there 

That said, I see a very short delay between when jenkins saw the build configs and created the watches,
and when the build was started.

we start watching the namespace here:

```
Apr 16, 2018 3:22:59 PM io.fabric8.jenkins.openshiftsync.BuildWatcher$1 doRun
INFO: creating Build watch for namespace ane-playground and resource version 27340071

```

we start monitoring the job run for the build that was created here:

```
Apr 16, 2018 3:23:10 PM io.fabric8.jenkins.openshiftsync.BuildSyncRunListener onStarted
INFO: starting polling build job/ane-playground/job/ane-playground-example-pipeline/1/
```

Only a 10 second difference, where some start up activity was still occurring in between that.

I see a couple more jobs start a few minutes later, but I don't have enough context to determine if a delay ocurred in starting them.

So, @andreaseger , going to need some more data here ... let's not assume anything 

a) how are you creating the pipeline?  how are you starting the builds? 

b) assuming you are using `oc start-build ...`  can you run `oc get <build name> -o yaml` right after you start the build 
Or, if you are starting from the jenkins console, run `oc get build ...` after you start it that way

c) can you go into the jenkins console and create a logger for `io.fabric8.jenkins.openshiftsync.BuildWatcher` and trace all events .... if so, reproduce with, and capture thos logs along with the jenkins pod logs 

thanks

**Comment by andreaseger** [comment]
So I had some time for this again today, so I redeployed the jenkins container(to have a fresh/empty log). It's setup with 2 namespaces in the order `jenkins ane-playground`. The pipeline in question is in the namespace `ane-playground`

After it started up I added a new log recorder via the web interface (is that what you meant with c) ? )

- 1148 UTC I clicked inside jenkins on "Build Now" (the log shows nothing till ~1150)
- 1150 UTC the build is now finally enqueued and started inside jenkins and also shows up in Openshift
- 1153 UTC I start a build in openshift via `oc start-build example-pipeline`. The OS console shows the build immediately
```yaml
apiVersion: v1
kind: Build
metadata:
  annotations:
    openshift.io/build-config.name: example-pipeline
    openshift.io/build.number: "27"
  creationTimestamp: 2018-04-18T11:53:34Z
  labels:
    build: example-pipeline
    buildconfig: example-pipeline
    openshift.io/build-config.name: example-pipeline
    openshift.io/build.start-policy: Serial
  name: example-pipeline-27
  namespace: ane-playground
  ownerReferences:
  - apiVersion: build.openshift.io/v1
    controller: true
    kind: BuildConfig
    name: example-pipeline
    uid: 6c61e75c-2dd0-11e8-bf6c-b838614683bd
  resourceVersion: "27565759"
  selfLink: /oapi/v1/namespaces/ane-playground/builds/example-pipeline-27
  uid: 1f8707a6-42ff-11e8-b1c5-b838614683bd
spec:
  nodeSelector: {}
  output: {}
  postCommit: {}
  resources: {}
  serviceAccount: builder
  source:
    contextDir: example
    git:
      ref: master
      uri: ssh://git@stash.runtastic.com:7999/os/image-jenkins.git
    sourceSecret:
      name: stash-runtastic-com
    type: Git
  strategy:
    jenkinsPipelineStrategy:
      jenkinsfilePath: Jenkinsfile
    type: JenkinsPipeline
  triggeredBy:
  - message: Manually triggered
status:
  config:
    kind: BuildConfig
    name: example-pipeline
    namespace: ane-playground
  output: {}
  phase: New
```
- 1155 UTC only now jenkins starts doing something `io.fabric8.jenkins.openshiftsync.CredentialsUtils upsertCredential`
- build is now running in jenkins and is synced as expected to OS


- 1159 UTC I changed the namespace order to `ane-playground jenkins`
- 1200 UTC starting a new build via `oc start-build` which immediately starts in jenkins
[pod.log](https://github.com/openshift/jenkins-sync-plugin/files/1923999/pod.log)

**Comment by gabemontero** [comment]
Thanks for the detail @andreaseger ... with it in hand, I believe I've been able to reproduce at least your flavor of the problem locally.

Just in case you were unaware, as  short term workaround, we previously released a way for users to control the polling interval.  You can now reduced that 5 minute default if you want to expedite your build processing in the meantime.

**Comment by andreaseger** [comment]
Great to hear that the infos helped.

I think for now we ca live with changing the order so that the namespace with the build_config/pipelines is the first namespace. So I guess we only would see the delay when adding new jenkins-slaves imagestreams in the second namespace which would be acceptable.

**Comment by gabemontero** [comment]
quick heads up @andreaseger , at least for me, if I switch things around enough times, I see either order fail.

I believe there is timing involved.

After tracing io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager I discovered that only 5 of the 10 potential watches we would create for 10 projects were completing successfully.

I then searched on okhttp3 a bit and found https://github.com/square/okhttp/issues/3801

And sure enough, when I look at the okhttp3.Dispatcher class that the fabric k8s client pulls in, it sets the conns per host to 5.

I strongly suspect we are hitting this limit.  I am not hunting throug the k8s client code to see if we can call the setter there and override that value.

If not, we'll need to work with the openshift.io team to update fabric so we can override (though we have some other fixes we are waiting on that we need from them before we can bump versions).

@bparees @coreydaley fyi ^^^
